File: dsa/6.hash_table/974_Subarray_Sums_Divisible_by_K.java
package hash_table;
import java.util.*;

class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        Map<Integer, Integer> mapVal = new HashMap<Integer, Integer>();
        int result = 0;
        int sumRemainder = 0;

        for (int i = 0; i < nums.length; i++){
            // to calculate correctly remainder with negative number
            sumRemainder = ((sumRemainder + nums[i]) % k + k) %k;

            if(sumRemainder % k == 0){
                result += 1;
            }

            if (mapVal.containsKey(sumRemainder) == true){
                result += mapVal.get(sumRemainder);
                mapVal.replace(sumRemainder, mapVal.get(sumRemainder) + 1);
            }
            else{
                mapVal.put(sumRemainder, 1);
            }
        }
        return result;
    }
}

File: dsa/6.hash_table/974_Subarray_Sums_Divisible_by_K.py
import typing as t

class Solution:
    def subarraysDivByK(self, nums: t.List[int], k: int) -> int:
        acc_sum: int = 0
        result: int = 0
        sum_mod_dict: t.Dict[int] = {
            0: 1
        }
        for ind in range(0, len(nums), 1):
            acc_sum += nums[ind]
            cur_mod: int = acc_sum % k
            if cur_mod in sum_mod_dict.keys():
                result += sum_mod_dict.get(cur_mod)
            sum_mod_dict[cur_mod] = sum_mod_dict.get(cur_mod, 0) + 1
        return result